vortex-scalar/src/primitive.rs

```diff
@@ -270,3 +272,47 @@ impl From<usize> for Scalar {
         Scalar::primitive(value as u64, Nullability::NonNullable)
     }
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum NumericOperator {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    // Missing from arrow-rs:
+    // Min,
+    // Max,
+    // Sqrt,
+    // Pow,
+    // In IEEE-754 operations involving NaN are non-commutative wrt NaN payload/quietness
+    // RAdd,
+    // RSub,
+    // RMul,
+    // RDiv,
+}
+
+impl PrimitiveScalar<'_> {
+    pub fn numeric_operator(
+        self,
+        other: PrimitiveScalar<'_>,
+        op: NumericOperator,
+    ) -> VortexResult<Scalar> {
+        if !self.dtype().eq_ignore_nullability(other.dtype()) {
+            vortex_bail!("types must match: {} {}", self.dtype(), other.dtype());
+        }
+
+        Ok(match_each_native_ptype!(self.ptype(), |$P| {
+            let lhs = self.typed_value::<$P>();
+            let rhs = other.typed_value::<$P>();
+            let nullability = self.dtype().nullability();
+            lhs.zip(rhs).map(|(lhs, rhs)| {
+                match op {
+                    NumericOperator::Add => Scalar::primitive(lhs + rhs, nullability),
+                    NumericOperator::Sub => Scalar::primitive(lhs - rhs, nullability),
+                    NumericOperator::Mul => Scalar::primitive(lhs * rhs, nullability),
+                    NumericOperator::Div => Scalar::primitive(lhs / rhs, nullability),
+                }
+            }).unwrap_or_else(|| Scalar::null(self.dtype().clone()))
```

Comment:
Because you checked for dtype_eq_ignore_nullability, you may now end up trying to construct a non-nullable null scalar right?

Reply:
Now using `as_nullable()`.

Reply:
Are you?

Reply:
Yeah, the `address comments` commit [did it here](https://github.com/spiraldb/vortex/pull/1640/commits/4ebb7edad2b454874df0aced585c204bd71e73fd#diff-8f1cbe207a4076299ad7913ac7d7d0db5fd2e99d1a689f984fcedfd1b91208faR315-R333).